src/Crypto/RNCryptor/V3/Decrypt.hs

```diff
@@ -102,10 +107,17 @@ decryptBlock :: RNCryptorContext
              -> ByteString
              -> (RNCryptorContext, ByteString)
 decryptBlock ctx cipherText = 
-  let clearText  = decryptCBC (ctxCipher ctx) (aesIV_ . rncIV . ctxHeader $ ctx) cipherText
-      !sz        = B.length cipherText
-      !newHeader = (ctxHeader ctx) { rncIV = (B.drop (sz - 16) cipherText) }
-      in (ctx { ctxHeader = newHeader }, clearText)
+  let clearText   = decrypt_ (ctxCipher ctx) (rncIV . ctxHeader $ ctx) cipherText
+      !newHMACCtx = update (ctxHMACCtx ctx) cipherText
+      !sz         = B.length cipherText
+      !newHeader  = (ctxHeader ctx) { rncIV = (B.drop (sz - 16) cipherText) }
+      in (ctx { ctxHeader = newHeader, ctxHMACCtx = newHMACCtx }, clearText)
+
+--------------------------------------------------------------------------------
+-- TBD:  need strictness here to overcome lazy pull from zip results stopping at
+-- first False?
+consistentTimeEqual :: ByteString -> ByteString -> Bool
+consistentTimeEqual a b = and $ B.zipWith (==) a b
```

Comment:
--------------------------------------------------------------------------------
-- "A consistent time function needs to be clear on which parameter is secret and
-- which one is untrusted. Your complexity must always be proportional to the length
-- of the untrusted data, not the secret."
-- 
-- Below, untrusted == arrived in the message, secret == computed
--
consistentTimeEqual :: ByteString -> ByteString -> Bool
consistentTimeEqual untrusted secret =
  let initialResult = if B.length secret == B.length untrusted then 0 else 1 :: Word8
      secretCycle = cycle (B.unpack secret)
      xorResults = zipWith xor (B.unpack untrusted) secretCycle
  in
    0 == foldl' (.|.) initialResult xorResults

The compiler likes it ... and the unit test passes :)